src/workerd/util/sqlite.c++

```diff
@@ -0,0 +1,835 @@
+    KJ_ASSERT(error != SQLITE_MISUSE, "SQLite misused: " code, ##__VA_ARGS__); \
+    KJ_ASSERT(error == SQLITE_OK, "SQLite failed: " code, sqlite3_errmsg(db), ##__VA_ARGS__); \
+  } while (false);
+// Version of `SQLITE_CALL` that can be called after inspecting the error code, in case some codes
+// aren't really errors.
+
+namespace {
+
+void disposeSqlite(sqlite3_stmt* stmt) {
+  sqlite3_finalize(stmt);
+
+  // Note that any returned error code is actually the last error to occur while executing the
+  // statement. This does not really mean that finalization failed, and the error in question
+  // should have been checked and reported earlier. So, we ignore it here.
+}
+
+template <typename T>
+class SqliteDisposer: public kj::Disposer {
+public:
+  void disposeImpl(void* pointer) const override {
+    disposeSqlite(reinterpret_cast<T*>(pointer));
+  }
+};
+
+template <typename T>
+kj::Own<T> ownSqlite(T* obj) {
+  static const SqliteDisposer<T> disposer;
+  return kj::Own<T>(obj, disposer);
+}
+
+}  // namespace
+
+// =======================================================================================
+
+Sqlite::Sqlite(Vfs& vfs, kj::PathPtr path) {
+  SQLITE_CALL_NODB(sqlite3_open_v2(path.toString().cStr(), &db,
+                                   SQLITE_OPEN_READONLY, vfs.getName().cStr()));
+}
+
+Sqlite::Sqlite(Vfs& vfs, kj::PathPtr path, kj::WriteMode mode) {
+  int flags = SQLITE_OPEN_READWRITE;
+  if (kj::has(mode, kj::WriteMode::CREATE)) {
+    flags |= SQLITE_OPEN_CREATE;
+
+    if (kj::has(mode, kj::WriteMode::CREATE_PARENT) && path.size() > 1) {
+      // SQLite isn't going to try to create the parent directory so let's try to create it now.
+      vfs.directory.openSubdir(path.parent(),
+          kj::WriteMode::CREATE | kj::WriteMode::MODIFY | kj::WriteMode::CREATE_PARENT);
+    }
+  }
+  KJ_REQUIRE(kj::has(mode, kj::WriteMode::MODIFY), "SQLite doesn't support create-exclusive mode");
+
+  SQLITE_CALL_NODB(sqlite3_open_v2(path.toString().cStr(), &db,
+                                   flags, vfs.getName().cStr()));
+}
+
+Sqlite::~Sqlite() noexcept(false) {
+  auto err = sqlite3_close(db);
+  if (err == SQLITE_BUSY) {
+    KJ_LOG(ERROR, "sqlite database destroyed while dependent objects still exist");
+    // SQLite actually provides a lazy-close API which we might as well use here instead of leaking
+    // memory.
+    err = sqlite3_close_v2(db);
+  }
+
+  KJ_REQUIRE(err == SQLITE_OK, sqlite3_errstr(err)) { break; }
+}
+
+static kj::Own<sqlite3_stmt> prepareSql(sqlite3* db, kj::StringPtr sqlCode, uint prepFlags) {
+  sqlite3_stmt* result;
+  const char* tail;
+
+  SQLITE_CALL(sqlite3_prepare_v3(db, sqlCode.begin(), sqlCode.size(), prepFlags, &result, &tail));
+  KJ_REQUIRE(result != nullptr, "SQL code did not contain a statement", sqlCode);
+  auto ownResult = ownSqlite(result);
+
+  while (*tail == ' ' || *tail == '\n') ++tail;
+  KJ_REQUIRE(tail == sqlCode.end(), "sqlite statement had leftover code", tail);
```

Comment:
As currently written, if we didn't have this require, than any leftover code would just be ignored. So if you did `Sqlite::run()` and gave it a string with two statements (with a semicolon separating them), only the first one would actually run.

I think what I should do instead, though, is actually allow multiple statements and parse and execute them all. It's a little weird, though, when bindings are present or results are returned. I think I should say that if there are multiple statements, only the last one can have any bindings and only the last statement's results will be enumerated by the returned `Query` object.

`Sqlite::prepare()` should still only accept a single statement because otherwise the code will get pretty weird.